internal/app/machined/pkg/controllers/runtime/cri_image_gc.go

```diff
@@ -248,12 +236,36 @@ func (ctrl *CRIImageGCController) cleanup(ctx context.Context, logger *zap.Logge
 				if expectedDigested, ok := expectedRef.(docker.Digested); ok && ref.Digest() == expectedDigested.Digest() {
 					// this is expected image by digest, inject other forms of the ref
 					expectedImageNames[digest] = struct{}{}
-					expectedImageNames[expectedRef.String()] = struct{}{}
+					expectedImageNames[expectedRef.Name()+"@"+digest] = struct{}{}
```

Comment:
is there some upstream code in containerd/docker that handles this, rather than us doing this?

Reply:
we're already using the parser part, so it's not completely hand-written

the actual set of images is more the way CRI containerd manages images